Feat/rpc endpoints to fetch data from key

hugocaillard opened the pull request:
### Description

Add two new endpoints:
- `/v2/clarity_marf_value/:clarity_marf_key`
- `/v2/clarity_metadata/:principal/:contract_name/:clarity_metadata_key`

I'm currently working on a Clarinet feature that allows to simulate running (or, said differently, to fork the mainnet state in the simnet data store). This requires the ability to fetch values from marf and metadata keys.

These new endpoints are similar to already existing endpoint (such as `getdatavar`,  `getmapentry`,  `getcontractsrc`, etc).

### Applicable issues

N/A

### Additional info (benefits, drawbacks, caveats)

Read more about this feature in the Clarinet issue: https://github.com/hirosystems/clarinet/issues/1503

I confirmed that the 2 new endpoints allow to achieve the desired goal by running a local devnet with these changes and forking the Clarinet simnet state from it. 

### Checklist

- [x] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)

jcnelson commented on clarity/src/vm/representations.rs:
```diff
@@ -51,6 +51,8 @@ lazy_static! {
         "({})|({})",
         *STANDARD_PRINCIPAL_REGEX_STRING, *CONTRACT_PRINCIPAL_REGEX_STRING
     );
+    static ref CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING: String =
```
This needs boundaries.  Unbounded regexes whose evaluation is accessible via the network are a DDoS bug waiting to happen.

hugocaillard replied:
All of the regex string I added end up being formatted together and exclusively used with boundaries ([here](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/stackslib/src/net/api/getclaritymarfvalue.rs#L63-L66) and [here](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/stackslib/src/net/api/getclaritymetadata.rs#L63-L68))

But since they are only used there, maybe I should build the regex more exclusively in these two files and not exposed in representation.rs. Would that be a better approach?

jcnelson commented on stackslib/src/net/api/getclaritymarfvalue.rs:
```diff
@@ -0,0 +1,218 @@
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r"^/v2/clarity_marf_value/(?P<clarity_marf_key>(vm-epoch::epoch-version)|({})|({}))$",
```
I strongly disagree with using the MARF key as input.  To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).  Instead, you should do the following:

* Accept the _hash_ of the key as input.  Then, you don't have to worry about verifying that it is well-formed, and you don't have to worry about explaining to the RPC API consumer how to construct it.  Furthermore, it's much safer and easier to determine if a string is a well-formed hash instead of a well-formed Clarity MARF key.
* Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

jcnelson replied:
Also, I dislike the path you chose, since it doesn't leave room for semantic expansion.  Instead, can you use the following:  `/v2/marf/clarity/{key-hash}`, where `key-hash` is the hash of the SIP-005 key.

hugocaillard replied:
In my use case (of Clarinet consuming this API), I already know the key (but not the hash). But I agree that it's not ideal to have the key as a use input.
Can you guide me to some code where I could see how to get the key <> key_hash conversion?

> Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

Not sure to follow here, where should I perform this change?

--- 

I think I just lack knowledge about this repo to fully address this comment, thanks for providing extra context 🙏

jcnelson replied:
Sure -- the (string) key's bytes are hashed via SHA512/256 to produce the MARF key.

> Not sure to follow here, where should I perform this change?

The trait `ClarityBackingStore` needs to be modified to add a method like `get_data_with_proof()`, but which takes an implementation of the `MarfTrieId` trait instead of a `&str` key.  There currently isn't a way to query the MARF via the ClarityDB with a bare hash.

obycode replied:
> I strongly disagree with using the MARF key as input. To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).

Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

Note that adding a new version of `get_data_with_proof` that takes in the `TriePath` would also require some significant changes to the `MemoryBackingStore`, since it currently directly uses the key to access the database.

jcnelson replied:
> Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

For the client, I think both approaches are equivalently difficult.  The hard part either way is constructing the key.  The difference between sending the key versus the hash of the key is trivial.

However, when comparing the difficulty of implementing the API (including changes to the Clarity DB) to only accept a hash versus the whole metadata key, the former is actually much easier.  First, the Clarity DB already performs lookups into the MARF by hash (albeit via a few layers of indirection), so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash` and adding a MARF method to lookup by `TrieHash`.  Second, by accepting a key, we would commit the API to validating that key so as to property return HTTP 400 responses [1] if the key was semantically invalid.  Adding all of the validation logic and test coverage for it would vastly exceed the work required to support lookups by `TrieHash`.

[1] I'm not a fan of returning a blanket 404 for invalid input.  A 404 specifically means that the resource could not be found, and doesn't tell the user what the actually need to know.  By contrast, any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash.

obycode replied:
Sure, I understand that argument. Beware of the changes needed to the `MemoryBackingStore` to support lookups by `TrieHash` -- this PR will unfortunately need to grow significantly. This will be more work than we would hope, but that's probably okay, since changes there are not affecting critical code.

hugocaillard replied:
It's a bit hard to follow but I've some questions based on my understading; 

>  "any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash"

We would return a 404 if the hash encodes an invalid key right? So the difference for the client seems to be quite the same to me. It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

> "so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash`"

I'm actually struggling to implement it in the MemoryBackingStore (around [here](https://github.com/stacks-network/stacks-core/blob/24566459e5a8bdb399e0fe2efbfa0dadc571c851/clarity/src/vm/database/sqlite.rs#L323-L329)), because I don't have access `TrieHash::from_data` here.
I don't see how to get from the key to the Hash.
Maybe it's trivial but I just don't know the repository good enough.

jcnelson replied:
> We would return a 404 if the hash encodes an invalid key right?

If this API endpoint only took a hash, then the only application responses to expect are 200 (with the serialized data) or 404 (if the hash does not map to a value).  The question of whether or not the hash corresponds to a semantically-valid but absent key or a semantically-invalid key does not need to be answered by the API endpoint.

> It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

That's the thing -- I think it's plausible that this makes a significant difference to the client.  If the client is an application that loads Clarity data, then the difference between receiving a 400 versus a 404 is that the former requires a bugfix (the client made an invalid key) and would likely result in the user being presented with an error condition to report, whereas the latter indicates that the submitted key is valid but not mapped.  Treating both as 404's would lead to a poorer user experience, since users would receive false negatives -- a bug in the application's key construction would be treated the same as the user submitting a query for nonexistent data.

jcnelson replied:
I'm happy to add the requisite Clarity DB patches to this PR, once I'm done helping @obycode with https://github.com/stacks-network/stacks-core/pull/5420.

jcnelson replied:
I went ahead and added the requisite ClarityDB functionality, and plumbed it through into this RPC endpoint.

hugocaillard replied:
Thanks a lot @jcnelson, this is super super helpful.
I'm currenlty working on addressing your other remaining comments.
Marking this one a resolved

jcnelson commented on stackslib/src/net/api/getclaritymetadata.rs:
```diff
@@ -0,0 +1,217 @@
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r"^/v2/clarity_metadata/(?P<address>{})/(?P<contract>{})/(?P<clarity_metadata_key>(analysis)|({}))$",
```
Please use `/v2/clarity/metadata/{principal}/{contract-name}/{clarity_metadata_key}` as your path.

jcnelson commented on stackslib/src/net/api/getclaritymetadata.rs:
```diff
@@ -0,0 +1,217 @@
+        let contract_identifier = self.contract_identifier.take().ok_or(NetError::SendError(
+            "`contract_identifier` not set".to_string(),
+        ))?;
+        let clarity_metadata_key = self.clarity_metadata_key.take().ok_or(NetError::SendError(
```
Clarity metadata keys have a well-defined structure, so the RPC endpoint should return HTTP 400 if the metadata key is ill-formed.  You will need to expand this method (or `try_parse_request` -- up to you) to validate the structure of the Clarity metadata key, including determining whether or not the key's `StoreType` and `var_name` values are supported values.  `var_name` will, unfortunately, take some effort because the Clarity DB codebase uses bare string literals in its calls to `ClarityDatabase::make_metadata_key()` instead of enums, so your PR should address this as well.

hugocaillard replied:
Apparently the metadata `var_name` can be an arbitrary string, where here it's a [variable_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1329), a [map_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1467), a [token_name](https://github.com/stacks-network/stacks-core/blob/26c448727def84bbef34ba4d14ec4fd5244a2390/clarity/src/vm/database/clarity_db.rs#L1811)).

Can we really use an enum here?

jcnelson replied:
Yes; however, some reserved `var_name` strings can only be paired with certain `StoreType`s.  For example, the use of `StoreType::Contract` would require `var_name` to be `contract-size`, `contract-src`, `contract-data-size`, or `contract`, but nothing else.

One way to implement these constraints could be to implement an enum to capture all of the valid `StoreType` / `var-name` pairings, and permit only the `StoreType` variants which allow arbitrary `var_name` values to have arbitrary `var_name` values.

jcnelson commented on stackslib/src/net/httpcore.rs:
```diff
@@ -247,6 +247,17 @@ pub mod request {
     }
 
+    /// Get a clarity key (MARF or Metadata) from a path's captures, given the name of the regex field.
+    pub fn get_clarity_key(captures: &Captures, clarity_key: &str) -> Result<String, HttpError> {
```
Because this is specific to your handler, you should implement it there instead.

jcnelson commented on clarity/src/vm/representations.rs:
```diff
@@ -64,6 +66,19 @@ lazy_static! {
+        *MARF_KEY_FOR_TRIP_REGEX_STRING,
+    );
+    pub static ref METADATA_KEY_REGEX_STRING: String = format!(
+        r"vm-metadata::\d+::(contract|contract-size|contract-src|contract-data-size|({}))",
```
You'll need boundaries here if you want this to be used in RPC calls.

jcnelson commented on clarity/src/vm/representations.rs:
```diff
@@ -64,6 +66,19 @@ lazy_static! {
+    pub static ref MARF_KEY_FOR_TRIP_REGEX_STRING: String = format!(
+        r"vm::{}::\d+::({})",
+        *CONTRACT_PRINCIPAL_REGEX_STRING,
+        *CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING,
```
You'll need boundiares here if you want this to be used in RPC calls.

jcnelson commented on stackslib/src/net/api/getclaritymetadata.rs:
```diff
@@ -0,0 +1,233 @@
+
+lazy_static! {
+    static ref CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING: String =
+        "[a-zA-Z]([a-zA-Z0-9]|[-_!?+<>=/*])*|[-+=/*]|[<>]=?".into();
```
This regex is very problematic because it matches strings of unbounded length, which opens a DoS vector on the node.

jcnelson replied:
Yes, because that fully-built regex does not impose boundaries either.  You need to limit the number of characters this regex can match.

hugocaillard replied:
Ooh right, sorry I missed that.
Fixed in my last commit

jcnelson commented on stackslib/src/net/api/getclaritymetadata.rs:
```diff
@@ -0,0 +1,233 @@
+    static ref CLARITY_NAME_NO_BOUNDARIES_REGEX_STRING: String =
+        "[a-zA-Z]([a-zA-Z0-9]|[-_!?+<>=/*])*|[-+=/*]|[<>]=?".into();
+    static ref METADATA_KEY_REGEX_STRING: String = format!(
+        r"vm-metadata::\d+::(contract|contract-size|contract-src|contract-data-size|({}))",
```
Same problem here with `\d+`

jcnelson replied:
Since the `\d+` must be only one of a handful of values, you should just match them explicitly.

jcnelson commented:
I went ahead and made the requisite changes to the MARF, `ClarityBackingStore` and all its implementations, and `ClarityDB` so you can load values by `TrieHash`.  I also removed `TriePath` from the codebase, since it's used exactly the same way that `TrieHash` is.

My comments on the getter for Clarity metadata still stand, however.  If we're going to take metadata keys as input, we need to validate them and return HTTP 400 if they're semantically invalid.

obycode commented on stacks-common/src/types/chainstate.rs:
```diff
@@ -30,6 +30,68 @@ impl_byte_array_serde!(TrieHash);
+
+    /// Convert to a String that can be used in e.g. sqlite
+    pub fn to_string(&self) -> String {
+        let s = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
+                          self.0[0],     self.0[1],       self.0[2],       self.0[3],
+                          self.0[4],     self.0[5],       self.0[6],       self.0[7],
+                          self.0[8],     self.0[9],       self.0[10],      self.0[11],
+                          self.0[12],    self.0[13],      self.0[14],      self.0[15],
+                          self.0[16],    self.0[17],      self.0[18],      self.0[19],
+                          self.0[20],    self.0[21],      self.0[22],      self.0[23],
+                          self.0[24],    self.0[25],      self.0[26],      self.0[27],
+                          self.0[28],    self.0[29],      self.0[30],      self.0[31]);
+        s
```
```rust
pub fn to_string(&self) -> String {
    self.0.iter().map(|byte| format!("{:02x}", byte)).collect()
}
```

jcnelson commented:
The reason we *didn't* do that is because that invokes the allocator too
much, per perf.

On Sat, Nov 9, 2024, 8:48 AM Brice Dobry ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In stacks-common/src/types/chainstate.rs
> <https://github.com/stacks-network/stacks-core/pull/4997#discussion_r1835377699>
> :
>
> > +        let s = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
> +                          self.0[0],     self.0[1],       self.0[2],       self.0[3],
> +                          self.0[4],     self.0[5],       self.0[6],       self.0[7],
> +                          self.0[8],     self.0[9],       self.0[10],      self.0[11],
> +                          self.0[12],    self.0[13],      self.0[14],      self.0[15],
> +                          self.0[16],    self.0[17],      self.0[18],      self.0[19],
> +                          self.0[20],    self.0[21],      self.0[22],      self.0[23],
> +                          self.0[24],    self.0[25],      self.0[26],      self.0[27],
> +                          self.0[28],    self.0[29],      self.0[30],      self.0[31]);
> +        s
>
> pub fn to_string(&self) -> String {
>     self.0.iter().map(|byte| format!("{:02x}", byte)).collect()}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-core/pull/4997#pullrequestreview-2425348048>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK4FVL5QOLAJUWV5NRDZ7YHEJAVCNFSM6AAAAABLKR7LICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIMRVGM2DQMBUHA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

obycode commented on docs/rpc/openapi.yaml:
```diff
@@ -486,6 +486,93 @@ paths:
+        - Smart Contracts
+      operationId: get_clarity_marf_value
+      description: |
+        Attempt to fetch the value of a MARF key. The key is identified with [Clarity MARF Key].
```
I think you can delete "The key is identified with [Clarity MARF Key]." since that is covered in the parameters.

obycode commented on docs/rpc/openapi.yaml:
```diff
@@ -486,6 +486,93 @@ paths:
+        - Smart Contracts
+      operationId: get_clarity_metadata_key
+      description: |
+        Attempt to fetch the metadata of a contract. The contract is identified with [Stacks Address] and [Contract Name] in the URL path. The metadata key is identified with [Clarity Metadata Key].
```
The names used should match, e.g. `[Stacks Address]` => `contract_address`.

obycode commented on stackslib/src/net/api/getclaritymetadata.rs:
```diff
@@ -0,0 +1,269 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2024 Stacks Open Internet Foundation
```
nit: for new files, I think they should just have:
```
// Copyright (C) 2024 Stacks Open Internet Foundation
```

obycode reviewed (approved):
LGTM

jcnelson commented on stackslib/src/net/api/tests/getclaritymetadata.rs:
```diff
@@ -0,0 +1,273 @@
+    assert_eq!(&preamble, request.preamble());
+
+    handler.restart();
+    assert!(handler.clarity_metadata_key.is_none());
```
Can you expand this test to cover error cases?  Specifically, can you verify that the user gets back HTTP 400 for invalid metadata key?

hugocaillard replied:
There are two test case for that below with `vm-metadata::2::contract-size` and `vm-metadata::9::contract-invalid-key`

jcnelson commented on stackslib/src/net/api/tests/getclaritymetadata.rs:
```diff
@@ -0,0 +1,273 @@
+    let expected = DataVariableMetadata {
+        value_type: TypeSignature::IntType,
+    };
+    assert_eq!(resp.data, expected.serialize());
```
Can you add test coverage here for HTTP 400 and HTTP 404 error cases?

hugocaillard replied:
Done

hugocaillard commented:
Thanks for the reviews @jcnelson 🙏 
I marked as resolved all of the addressed comment
Re-requested Brice review as well since it was dismissed, but he already approved earlier.

obycode reviewed (dismissed):
🎉

hugocaillard commented:
@jcnelson I'm actually seeing something weird with the last test I added.

If I add a request after this [call](https://github.com/stacks-network/stacks-core/blob/7f3191b6aa697881d9dc088832dea4a4d1497d14/stackslib/src/net/api/tests/getclaritymetadata.rs#L256-L264), or if I add a previous request hitting this [code path](https://github.com/stacks-network/stacks-core/blob/a94abfda195ca653b6344924eaa62fba7bf2e250/stackslib/src/net/api/getclaritymetadata.rs#L121-L141) (such as a request `vm-metadata::9::invalid-contract-metadata-key`, the following request don't work and [this assertion fails](https://github.com/stacks-network/stacks-core/blob/a94abfda195ca653b6344924eaa62fba7bf2e250/stackslib/src/net/api/tests/mod.rs#L1080).

Is it possible that this code path breaks the test runner?
It works just fine in practice when calling this endpoint

jcnelson commented:
@hugocaillard Can you push a test case so I can take a look?

jcnelson reviewed (changes requested):
Placing a hold on this until we get further clarity from @hugocaillard

hugocaillard commented on stackslib/src/net/api/tests/getclaritymarfvalue.rs:
```diff
@@ -0,0 +1,203 @@
+
+    let resp = response.decode_data_var_response().unwrap();
+    let balance = Value::try_deserialize_hex(&resp.data[2..], &TypeSignature::IntType, false);
+    assert_eq!(balance, Ok(Value::Int(256_000_000_000)));
```
I'm not sure that 256M is the right expected value 🤔

obycode replied:
Everything looks good to me, but I'm also unclear how this balance reaches 256,000 STX.

hugocaillard replied:
@jcnelson Does that make sense to you?

obycode reviewed (approved):
LGTM!

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
